Content.Shared/_Crescent/Hardpoints/SharedHardpointSystem.cs

```diff
@@ -104,15 +178,33 @@ public bool TryAnchorToAnyHardpoint(EntityUid uid, HardpointAnchorableOnlyCompon
                 continue;
             if (hardComp.CompatibleSizes < component.CompatibleSizes)
                 continue;
-            AnchorEntityToHardpoint(uid, entity, component, hardComp, gridUid.Value);
+            AnchorEntityToHardpoint(uid, entity, component, hardComp, gridUid.Value, user);
             return true;
         }
 
         return false;
     }
 
-    public void AnchorEntityToHardpoint(EntityUid target, EntityUid anchor,HardpointAnchorableOnlyComponent targetComp, HardpointComponent hardpoint, EntityUid grid)
+    public void AnchorEntityToHardpoint(EntityUid target, EntityUid anchor, HardpointAnchorableOnlyComponent targetComp, HardpointComponent hardpoint, EntityUid grid, EntityUid? user = null)
     {
+        // Double-check the limit right before anchoring (server-side safety)
+        if (!CanInstallOnGrid(grid))
+        {
+            var currentCount = GetCurrentHardpointCount(grid);
+            var maxAllowed = currentCount <= 1 ? 4 : currentCount + 2;
+
+ //           Logger.Warning($"[HardpointDebug] BLOCKING anchor attempt - limit would be exceeded on grid {grid}");
+
+            // Show popup to the user who tried to anchor, similar to anchorable system
+            if (user != null)
+            {
+                  _popup.PopupClient($"Cannot exceed grid limit of {maxAllowed} hardpoints!", anchor, user.Value);
+            }
+            return;
+        }
+
+//        Logger.Info($"[HardpointDebug] Anchoring entity {target} to hardpoint {anchor} on grid {grid}");
```

Comment:
hands off , hook onto the beforeAnchor event and cancel it before it even gets to the hardpoint system,